python/truelayer_signing/utils.py

```diff
@@ -0,0 +1,143 @@
+from __future__ import annotations
+
+# std imports
+import base64
+import json
+from enum import Enum
+from typing import Dict, Mapping, MutableMapping, Optional, Tuple, TypeVar
+
+T = TypeVar('T', bound='TlJwsBase')
+
+
+class HttpMethod(str, Enum):
+    POST = "POST"
+    GET = "GET"
+    PATCH = "PATCH"
+    PUT = "PUT"
+    DELETE = "DELETE"
+
+
+class TlJwsBase:
+    pkey: str
+    http_method: HttpMethod
+    path: str
+    headers: MutableMapping[str, str]
+    body: str
+
+    def __init__(
+        self,
+        pkey: str,
+        http_method: HttpMethod = HttpMethod.POST,
+        path: str = "",
+        headers: Optional[Dict[str, str]] = None,
+        body: str = ""
+    ):
+        self.pkey = pkey
+        self.http_method = http_method
+        self.path = path
+        self.headers = {} if headers is None else headers
+        self.body = body
+
+    def set_method(self: T, http_method: HttpMethod) -> T:
+        """
+        Add the request method, defaults to `"POST"` if unspecified.
+        """
+        self.http_method = http_method
+        return self
+
+    def set_path(self: T, path: str) -> T:
+        """
+        Add the request absolute path starting with a leading `/` and without
+        any trailing slashes.
+        """
+        self.path = path
+        return self
+
+    def add_headers(self: T, headers: MutableMapping[str, str]) -> T:
+        """
+        Appends multiple header names & values.
+
+        Warning: Only a single value per header name is supported.
+        """
+        for k, v in headers.items():
+            self.headers[k] = v
+        return self
+
+    def add_header(self: T, header: str, value: str) -> T:
+        """
+        Add a header name & value.
+        May be called multiple times to add multiple different headers.
+
+        Warning: Only a single value per header name is supported.
+        """
+        self.headers[header] = value
+        return self
+
+    def set_body(self: T, body: str) -> T:
+        """
+        Add the full request body.
+
+        Note: This **must** be identical to what is sent with the request.
+        """
+        self.body = body
+        return self
+
+
+def build_v2_signing_payload(
+    method: str,
+    path: str,
+    headers: Mapping[str, str],
+    body: str
+) -> str:
+    """
+    Build a TLv2 signing payload.
+    Retruns signing payload as a string
+
+    ### Example
+    ```txt
+    POST /test-signature
+    Idempotency-Key: 619410b3-b00c-406e-bb1b-2982f97edb8b
+    {"bar":123}
+    ```
+    """
+    payload = f"{method} {path}\n"
+    for key, value in headers.items():
+        payload += f"{key}: {value}\n"
+    payload += body
+    return payload
+
+
+def build_v2_jws_b64(
```

Comment:
We could catch and return exceptions wrapped in our tl exception type for json encoding errors. If I were using this API I would like to be sure that catching `TlSigningException` would handle expected errors like json and base64 encoding errors, what do you think?

Reply:
We should never get a JSON error unless the function is used improperly as this is an internal function I think it's a reasonable assumetion. I did add some documentation that a `UnicodeEncodeError` can be raised. 

What do you think? would it be worth wrapping the exception

Reply:
iirc in other libs we wrapped all such errors, it makes it easier for users verifying a signature to handle a single exception for any expected case, e.g. sending some nonsense signature header.

Reply:
Good point, I hadn't considered the verification route. Will set a custom error.